nselib/codesys3.lua

```diff
@@ -0,0 +1,252 @@
+    new = function(self, data)
+      local o = { data = data }
+
+      if ( not(data) or #data < 6 ) then
+        return false, "Response isn't long enough the be a Name Service Response"
+      end
+
+      local hopinfo, packetinfo, address_lengths, pos
+      o.magic, hopinfo, packetinfo, o.service_id, o.message_id, address_lengths, pos = string.unpack(">BBBBBB", data)
+
+      -- parse hopinfo field
+      o.hop_count = hopinfo >> 3
+      o.header_length = hopinfo & 7
+
+      -- parse packetinfo field
+      o.priority = packetinfo >> 6
+      o.signal = (packetinfo >> 5) & 1
+      o.address_type = (packetinfo >> 4) & 1
+      o.data_length = packetinfo & 0xf
+
+      -- sanity check a few fields to determine if this is a name service broadcast response
+      if o.magic ~= 0xc5 or o.service_id ~= CodesysV3.ServiceIDs.NameServiceResponse then
+        return false, "Response has the wring magic value or isn't a Name Service Response"
+      end
+
+      -- skip ahead to the address fields by skipping the length of the header
+      pos = o.header_length * 2
+
+      -- skip over the address fields using the address length fields in the header
+      pos = pos + (address_lengths & 0xf) * 2
+      pos = pos + (address_lengths >> 4) * 2
+
+      -- skip the padding bytes if necessary
+      if pos % 4 ~= 0 then
+        pos = pos + (pos % 4)
+      end
+
+      -- lua strings are 1-indexed, so adjust the position pointer...
+      pos = pos + 1
+      
+      -- at this point we are at the packet payload
+
+      -- parse the payload header
+      o.package_type, o.version, o.request_id, pos = string.unpack("<HHI", data, pos)
+
+      
+      -- check the package type we are expecting
+      if o.package_type ~= CodesysV3.NameServicePackageTypes.Identification then
+        return false, "The payload in the response isn't a Name Service Identification"
+      end
+
+      -- TODO: we should handle other versions than v4.00 as well
+      if o.version ~= 0x0400 then
+        return false, "The response contained a different version than v4.00 which we don't support right now"
+      end
+
+      stdnse.debug1("Received Codesys V3 Name Service Identification Response packet we can parse")
+
+      -- FIMXE: Intel vs Motorola Byte Order - means: Little vs Big endian. Does endianness of fields in the
+      --        packet later on have different endianness as well? Don't have a Big Endian device to test at hand
+
+      -- Start parsing the PLC identification payload
+      local parentAddrSize, nodeNameLength, deviceNameLength, vendorNameLength,
+            serialNumberLength, oemDataLength
+
+      o.maxChannels, o.intelByteOrder, o.addrDifference, parentAddrSize, pos = string.unpack("<I2 I1 I1 I2", data, pos)
+      nodeNameLength, deviceNameLength, vendorNameLength, pos = string.unpack("<I2 I2 I2", data, pos)
+      o.targetType, o.targetId, o.targetVersion, o.flags, pos = string.unpack("<I4 I4 I4 I4", data, pos)
```

Comment:
If I convert it into a string, we'd lose comparability. If we leave it in a DWORD, we can just use equals, lesser than or greater than and it would just work when comparing different version strings. Not that I need it right now, but maybe in the future?